onnx/defs/nn/defs.cc

```diff
@@ -292,6 +292,75 @@ computes the output.)DOC";
         "number of groups input channels and output channels are divided into, default is 1.",
         AttributeProto::INT,
         static_cast<int64_t>(1));
+    schema.TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        if (!hasNInputShapes(ctx, 1)) {
+          return;
+        }
+
+        // don't bother with legacy auto_pad for now
+        if (ctx.getAttribute("auto_pad")) {
+          return;
+        }
+
+        std::vector<int64_t> dilations;
+        if (!getRepeatedAttribute(ctx, "dilations", dilations)) {
+          for (int i = 0; i < ctx.getInputType(0)->tensor_type().shape().dim_size(); ++i) {
+            dilations.push_back(1);
+          }
+        }
+
+        std::vector<int64_t> kernel_shape;
+        if (!getRepeatedAttribute(ctx, "kernel_shape", kernel_shape)) {
+          for (int i = 2; i < ctx.getInputType(1)->tensor_type().shape().dim_size(); ++i) {
+            kernel_shape.push_back(ctx.getInputType(1)->tensor_type().shape().dim(i).dim_value());
+          }
+        }
+
+        std::vector<int64_t> pads;
+        if (!getRepeatedAttribute(ctx, "pads", pads)) {
+          for (int i = 0; i < ctx.getInputType(0)->tensor_type().shape().dim_size(); ++i) {
+            pads.push_back(0);
+            pads.push_back(0);
+          }
+        }
+
+        std::vector<int64_t> strides;
+        if (!getRepeatedAttribute(ctx, "strides", strides)) {
+          for (int i = 0; i < ctx.getInputType(0)->tensor_type().shape().dim_size(); ++i) {
+            strides.push_back(1);
+          }
+        }
+
+        *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape()->add_dim() =
+          ctx.getInputType(0)->tensor_type().shape().dim(0);
+        *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape()->add_dim() =
+          ctx.getInputType(1)->tensor_type().shape().dim(0);
+
+        for (size_t i = 0; i < kernel_shape.size(); ++i) {
+          if (!ctx.getInputType(0)->tensor_type().shape().dim(2 + (int)i).has_dim_value()) {
+            continue;
+          }
+          // how big is the input, including padding
+          int64_t effective_input_size = ctx.getInputType(0)->tensor_type().shape().dim(2 + (int)i).dim_value();
+          effective_input_size += pads[i];
+          effective_input_size += pads[i + kernel_shape.size()];
+
+          // accounting for dilation, how big is the kernel in this dimension
+          int64_t effective_kernel_size = kernel_shape[i];
+          effective_kernel_size = (effective_kernel_size - 1) * dilations[i] + 1;
+
+          // output size, no accounting for strides
+          int64_t kernel_positions = effective_input_size - effective_kernel_size + 1;
+
+          // output size, accounting for strides
+          int64_t effective_kernel_positions = kernel_positions / strides[i];
```

Comment:
The formula here seems incorrect:

For example: pad=0, dilation=1, input=5, kernel=5, stride=2, you will get output=0

I think right formula should be:
Output[i+2] = (Input[i+2] + padding[i] + padding[i+len(kernel_shape)] - dilations[i] *( kernel_shape[i] - 1) - 1) / stride[i] + 1